selfdrive/car/isotp_parallel_query.py

```diff
@@ -131,16 +134,24 @@ def get_data(self, timeout, total_timeout=None):
             results[tx_addr] = dat[len(expected_response):]
             request_done[tx_addr] = True
         else:
-          request_done[tx_addr] = True
-          cloudlog.warning(f"iso-tp query bad response: 0x{dat.hex()}")
+          error_code = dat[2] if len(dat) > 2 else -1
+          if error_code == 0x78:
+            pending_responses[tx_addr] = time.monotonic()
+            if self.debug:
+              cloudlog.warning(f"iso-tp query response pending: {hex(tx_addr)}")
+          else:
+            request_done[tx_addr] = True
+            cloudlog.warning(f"iso-tp query bad response: 0x{dat.hex()}")
 
       cur_time = time.monotonic()
-      if cur_time - last_response_time > timeout:
+      if (not len(pending_responses) and (cur_time - last_response_time > timeout)) or \
+        (len(pending_responses) and (cur_time - max(pending_responses.values()) > self.response_pending_timeout)):
         for tx_addr in msgs:
-          if (request_counter[tx_addr] > 0) and (not request_done[tx_addr]):
+          if (request_counter[tx_addr] > 0 or pending_responses[tx_addr] != -1) and (not request_done[tx_addr]):
```

Comment:
```suggestion
          if (request_counter[tx_addr] > 0 or pending_responses[tx_addr] != 0) and (not request_done[tx_addr]):
```

This and the defaultdict are needed to avoid KeyError exceptions for queries without a timer running.

https://termbin.com/pw7h